streams/lms_metadata.py

```diff
@@ -0,0 +1,209 @@
+    self.player_name = name
+    self.server = server
+    self.port = port
+    self.ip = ip # Generally populated later
+    self.meta_ref_rate = meta_ref
+    self.debug = debug
+    self.meta = MetadataHolder(artist = 'Loading...',
+                          album = 'Loading...',
+                          track = 'Loading...',
+                          image_url = 'static/imgs/lms.png')
+    if not server:
+      self.meta.album =  'If loading takes a long time,'
+      self.meta.track = 'consider adding hostname to stream config'
+
+
+  def flatten(self, lms_info: dict) -> dict:
+    """ LMS returns data in a very verbose piecewise format to avoid name collisions,
+          this makes it easier to use, by disregarding possible collisions"""
+    x = 0
+    flat_data = {}
+    for item in lms_info:
+      for info in item:
+        flat_data[info] = lms_info[x][info]
+      x += 1
+    return flat_data
+
+  def connect(self):
+    """Discovers LMS Player and then requests metadata repetitively"""
+    connected = False
+    self.meta.save_file(self.player_name)
+
+    # When not connected, search for player to connect to by the proper name
+    attempt_resolution = True # Attempt by name only once, as a common error is a nonresolvable DNS hostname
+    abort = False # Used to stop looping without initiating the secondary loop
+    while not connected and not abort:
+      try:
+        if self.server is not None and self.ip is None and attempt_resolution is True:
+          host = socket.gethostbyname(self.server)
+          logging.debug(f"hostname: {self.server}")
+          logging.debug(f"ip: {host}")
+          self.ip = host
+      except Exception:
+        logging.warning("Could not resolve hostname, trying to find LMS server manually")
+        logging.warning("Set local network DNS settings to avoid this issue in the future")
+        attempt_resolution = False
+
+      try:
+        if self.ip is None:
+          machine = platform.machine()
+          logging.debug(f"platform.machine() output: {machine}")
+
+          find_lms_server = None
+          if machine == "x86_64":
+            find_lms_server = "bin/x64/find_lms_server"
+          elif machine == "armv7l":
+            find_lms_server = "bin/arm/find_lms_server"
+          else:
+            self.meta.artist = "Unsupported CPU architecture for LMS MetaData"
+            self.meta.album = "Please contact AmpliPi Support:"
+            self.meta.track = "support@micro-nova.com"
+            logging.error("LMS metadata reader has detected an unsupported chipset")
+            abort = True
+
+          # Much faster method of connecting to the metadata server using code from: https://github.com/ralph-irving/squeezelite/blob/master/tools/find_server.c
+          ip_find = subprocess.run([find_lms_server], check=True, capture_output=True, text=True)
```

Comment:
what is the "less fast" way of connecting to the metadata server? is it an acceptable fallback for the unsupported CPU arch case? ie, is it pure Python?

Reply:
The less fast way is via manually scraping the whole network with python nmap; one that was prohibitively slow and also annoyed the hell out of every other device on the network by going door to door and asking "Are you an LMS server?"

Technically possible, doesn't seem worthwhile or sensible

IMO it'd be better to implement the ability to write down the IP as this script already supports that and is simply awaiting the LMS stream to add it in as an option during stream creation

Reply:
> IMO it'd be better to implement the ability to [input] the IP

after reading through the rest of this, I tend to agree. if you:
a) haven't provided a hostname or IP
b) the LMS server cannot be autodiscovered, with the binary provided
then you've already come to the end of the road. making (a) accept more inputs seems like a reasonable path here